### PR TITLE
add js api behaviour tests

### PR DESCRIPTION
**Pull request by michaeladda**
dependent on binaris/bolt#96

### CODE REVIEW

File: test/jsApi.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    -   in: bn remove returnInput
+- test: Test js method 
+  steps:
+    -   in: bn create node8 returnMethod
+    -   in: |-
+            echo 'exports.handler = (_, req) => req.method;' >function.js
+    -   in: bn deploy returnMethod
+    -   in: curl -s -X GET "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+        out: '"GET"'
+#    -   in: curl -s -X HEAD "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
```
Comment (arielshaqed) [inline test/jsApi.yml]: If you place code in comments please explain in the comments why they contain code.

File: test/spec.yml [hunk 2]
```diff
@@ -467,7 +467,7 @@
             some error
             Error: some error
                 at Bolt.exports.handler [as userEntryPoint] (/code/function.js:2:9)
-                at Bolt.run (/opt/binaris/dist/bolt.js:62:31)
+                at *
```
Comment (arielshaqed) [inline test/spec.yml]: Yay!

File: test/jsApi.yml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#        out: '"HEAD"'
+    -   in: curl -s -X POST "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+        out: '"POST"'
+    -   in: curl -s -X PUT "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+        out: '"PUT"'
+    -   in: curl -s -X DELETE "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+        out: '"DELETE"'
+#    -   in: curl -s -X CONNECT "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
```
Comment (rylandg) [inline test/jsApi.yml]: Comments are confusing

File: test/jsApi.yml [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+#        out: '"CONNECT"'
+    -   in: curl -s -X OPTIONS "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+        out: '"OPTIONS"'
+#    -   in: curl -s -X TRACE "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+#        out: '"TRACE"'
+    -   in: curl -s -X PATCH "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnMethod"
+        out: '"PATCH"'
+    -   in: bn remove returnMethod
+- test: Test js data input
```
Comment (rylandg) [inline test/jsApi.yml]: Nit: (but please fix) I really prefer a newline in between different test blocks. It's much easier to read when they are separated.

File: test/jsApi.yml [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    -   in: |-
+            echo 'exports.handler = (input) => ({ inputType: typeof input, input});' >function.js
+    -   in: bn deploy returnInput
+    -   in: curl -s "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnInput"
+        out: '{"inputType":"object","input":{}}'
+    -   in: curl -s --data '{"' "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnInput"
+        out: '{"inputType":"undefined"}'
+    -   in: curl -s --data '{}' "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnInput"
+        out: '{"inputType":"object","input":{}}'
+    -   in: curl -s --data '{"foo":"bar","yet":123}' "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnInput"
```
Comment (rylandg) [inline test/jsApi.yml]: nit: might be cleaner to alias `https://${BINARIS_INVOKE_ENDPOINT}/v1/run/` as `BINARIS_RUN_ENDPOINT`.

File: test/jsApi.yml (hunk 1 above)
Comment (michaeladda) [reply test/jsApi.yml]: removed for now. will put it back once it works

File: test/jsApi.yml [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    -   in: bn remove returnReqBody
+- test: Test js headers input
+  steps:
+    -   in: bn create node8 returnReqTestHeader
+    -   in: |-
+            echo 'exports.handler = (_, req) => req.headers.test;' >function.js
+    -   in: bn deploy returnReqTestHeader
+    -   in: curl -s "https://${BINARIS_INVOKE_ENDPOINT}/v1/run/${BINARIS_API_KEY}/returnReqTestHeader"
+        out: ''
+    -   in: |-
```
Comment (rylandg) [inline test/jsApi.yml]: Mostly just a preference question. Why use `|-` operator instead of single quoting the curl statements?

**Comment by rylandg** [comment]
@arielshaqed the verbosity or number of tests might be off but I think these are definitely important. Any functionality that we directly surface to the user, should have an accompanying test at the user level. We aren't testing that bolt can now handle headers with these tests but rather that headers work e2e through the system.

**Comment by michaeladda** [comment]
@arielshaqed I'm merging, but I'd love to discuss this f2f tomorrow and address any concern. This PR is important. I don't want to miss anything. Thanks!

**Pull request merged**